### ISSUE TITLE
If you rename a global variable in a child scope, the global variable is removed

### ISSUE THREAD

**Issue by KirkMunro** [PowerShell/PowerShell#10302]
When using the `Rename-Item` cmdlet on an existing provider item, I reasonably expect that cmdlet to rename the item that I am referencing. In the variable provider, if the variable that you are referencing is not defined in the current scope (e.g. if it is a global variable and you are in a child scope), renaming that variable via `Rename-Item` doesn't actually rename the variable in the scope it is in. The variable gets renamed, but it is moved to the child scope instead of remaining in the scope in which it was defined.

This behavior is akin to renaming a file from a location other than the folder where it is stored, and having the file moved to your current folder as part of the rename operation. It is unexpected and should be fixed.

# Steps to reproduce

```powershell
# Show the global ErrorActionPreference variable
Get-Variable -Name ErrorActionPreference -Scope Global

# In a child scope, rename that variable to 'foo'
& {
    Rename-Item variable:ErrorActionPreference -NewName foo
}

# Show the value of the global foo variable (this should exist)
Get-Variable -Name foo -Scope Global

# Show the value of the global ErrorActionPreference variable (this should not exist)
Get-Variable -Name ErrorActionPreference -Scope Global
```

# Expected behavior

```none
Name                           Value
----                           -----
ErrorActionPreference          Continue
foo                            Continue

Get-Variable : Cannot find a variable with the name 'ErrorActionPreference'.
At line:13 char:1
+ Get-Variable -Name ErrorActionPreference -Scope Global
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (ErrorActionPreference:String) [Get-Variable], ItemNotFoundException
+ FullyQualifiedErrorId : VariableNotFound,Microsoft.PowerShell.Commands.GetVariableCommand
```

# Actual behavior

```none
Name                           Value
----                           -----
ErrorActionPreference          Continue

Get-Variable : Cannot find a variable with the name 'foo'.
At line:10 char:1
+ Get-Variable -Name foo -Scope Global
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (foo:String) [Get-Variable], ItemNotFoundException
+ FullyQualifiedErrorId : VariableNotFound,Microsoft.PowerShell.Commands.GetVariableCommand

Get-Variable : Cannot find a variable with the name 'ErrorActionPreference'.
At line:13 char:1
+ Get-Variable -Name ErrorActionPreference -Scope Global
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (ErrorActionPreference:String) [Get-Variable], ItemNotFoundException
+ FullyQualifiedErrorId : VariableNotFound,Microsoft.PowerShell.Commands.GetVariableCommand
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.2.1
PSEdition                      Core
GitCommitId                    6.2.1
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
Is it specific for `ErrorActionPreference` name?

**Comment by KirkMunro** [comment]
No, not at all. That was just what I was using when I encountered the error. Here's a screenshot showing it with an arbitrary variable that I create:

![image](https://user-images.githubusercontent.com/3515485/62545039-4ad9a380-b837-11e9-8526-92bebbef3e52.png)

**Comment by KirkMunro** [comment]
For reference, without debugging (just reading the code), [this block of code](https://github.com/PowerShell/PowerShell/blob/1be3f4cc0e465ae11ad8e59e9060f5a59e4762eb/src/System.Management.Automation/namespaces/SessionStateProviderBase.cs#L941-L944) seems to be why this happens. Other than the `FileSystem` and `Registry` providers, all other providers that ship with PowerShell seem to take the same create-new-item-and-then-remove-old-item approach to renaming.

Sure enough, here's the Alias provider showing the same issue:

```none
PS C:\> New-Alias test Get-Date -Force
>> Get-Alias test
>> & {
>> Rename-Item Alias:test -NewName date
>> }
>> Get-Alias date
>> Get-Alias test

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           test -> Get-Date

Get-Alias : This command cannot find a matching alias because an alias with the name 'date' does not exist.
At line:6 char:1
+ Get-Alias date
+ ~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (date:String) [Get-Alias], ItemNotFoundException
+ FullyQualifiedErrorId : ItemNotFoundException,Microsoft.PowerShell.Commands.GetAliasCommand

Get-Alias : This command cannot find a matching alias because an alias with the name 'test' does not exist.
At line:7 char:1
+ Get-Alias test
+ ~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (test:String) [Get-Alias], ItemNotFoundException
+ FullyQualifiedErrorId : ItemNotFoundException,Microsoft.PowerShell.Commands.GetAliasCommand
```

If we're to fix this, there should probably be a `ScopedProviderBase` intermediate class that derives from `SessionStateProviderBase` and that we consider deriving `AliasProvider`, `FunctionProvider`, and `VariableProvider` from that intermediate class to ensure that scope is persisted for any update operations that demonstrate buggy/unexpected behavior like this. i.e. if `Rename-Item`, `Set-Item`, or other item update commands are only supported in the current scope (probably the only thing we can support at this point while maintaining backwards compatibility where it makes sense to do so), then:

1. Those operations should create the new item in the current scope and leave the original item in the parent scope.
1. Documentation should clearly indicate how these commands work with respect to scope.

**Comment by iSazonov** [comment]
I'd expect too that for PowerShell it would be more "native" doesn't remove the global/high-level-scope variable and create variable in local scope, using explicit scope prefix is more safe. It seems we do such things when we create a new scope with copying variables into it.

**Comment by KirkMunro** [comment]
Also for context, I discovered this issue while working on some Pester tests that remove global variables only to then re-create them afterward. When they re-create, they weren't setting the attributes on those variables properly. Instead, they were just assigning a value. That risks impacting other tests, so I thought a better approach would be to rename the variable in the global scope, define a new global variable that I want for my test, and then afterward remove the global variable I defined and then rename the original variable back in the global scope. That way the state of the original variable is preserved beyond the test where it is changed or hidden. I would like that to be possible with `Rename-Item`, but today it is not.

I'm hoping we can address this without being so conservative that it still isn't possible. If I rename a variable, I would like it to be renamed in the scope in which it resides -- that should be possible given that I can manipulate variables in any scope using PowerShell already. The main concern is doing so without breaking something unintentionally (although I think we should at least break compatibility with the broken way this works now since it is just a source of bugs/unexpected behavior).
